### PR TITLE
DRILL-7620: Fix plugin mutability issues

### PR DESCRIPTION
**Pull request by paul-rogers**
# [DRILL-7620](https://issues.apache.org/jira/browse/DRILL-7620): DRILL-7620: Fix plugin mutability issues

A recent commit made the plugin registry more strict about the rule that, once a plugin is registered, it must be immutable. A flaw enforcing that rule in the UI put the registry in an inconsistent state.

## Description

The plugin registry maintains two maps: one from name to config, the other from config to instance. There is also a third: from config to instance, used for "ephemeral" plugins: those that appear in query plans but are not named.

The config-to-instance map requires that a plugin be immutable once placed in the registry. Yet, as it turns out, we have quite a bit of code that changes plugins that reside in the registry. The UI did this: obtain a plugin, set the plugin to enabled/disabled, then update the registry. When this happens, the registry gets out of sync: the state of the plugin has already changed outside the registry, leaving the registry in the wrong state.

Unit tests sometimes grab a file system config and add on a few format configs. Again, this makes the config in the registry inconsistent with its map entry.

To help solve this problem, added methods to the registry to:

* Enable/disable plugins by name, using the correct steps.
* Copy a plugin for edit to discourage grabbing a config and changing it.
* Properly add/remove format plugins for the file system config.

A future PR may make plugin configs immutable. For now, we all must be on our best behavior not to change stored configs.

### Error Reporting

Error reporting in the registry was a bit awkward with everything reported by a `ExecutionSetupException`. Added exceptions specific to the registry to clearly report user-caused errors (undefined plugin names) vs. more subtle issues (serialization issues.) Changed all clients to use the new exceptions.

### UI

The UI (`StorageResources`) did more work than it really should, serializing and deserializing configs more than needed. Streamlined this code to use methods in the registry. An advantage is that code in the registry is easier to test than code in the REST server.

The UI allows users to include comments in their JSON, but we, of course, immediately discard those comments. To avoid the UI doing its own deserialization, moved that logic into the registry, and enhanced the `LogicalPlanPersistence` mapper to allow comments.

The UI allows users to export plugins as HOCON. But, Jackson cannot write objects as HOCON. Drill is promising something it can't deliver. Left that issue alone for now.

### Resolving Plugins from within Plugins

All group scans, sub scans and so on deserialize their plugins. There were several different ways to do this: some checked for a null return value, others did not. After adding the plugin exceptions, some operations that threw unchecked exceptions now throw checked exceptions. To avoid the need to copy-paste try blocks in every group and sub scan, added new methods to do the lookups, check for errors, and throw an unchecked exception if anything goes wrong.

### Clearer Cache/Storage Logic

The bug which @agozhiy found was due to a gap in synchronizing the in-memory plugin instance cache with the persistent config cache. Realized that the `put` operation did not even have to try to synchronize, we do that on `get` operations. After that fix, the described problem no longer appears. 

## Documentation

None.

## Testing

Added even more plugin registry unit tests. Made the tests more carefully honor the "immutable plugin rule." Manually performed the UI steps in the JIRA ticket.

### CODE REVIEW

File: exec/java-exec/src/test/java/org/apache/drill/exec/store/TestPluginRegistry.java [hunk 1]
```diff
@@ -195,36 +391,90 @@ public void testBasicLifecycle() throws Exception {
...
       } catch (UserException e) {
         // Expected
       }
+    }
+  }
+
+  @Test
+  public void testEphemeralWithoutInstance() throws Exception {
+    ClusterFixtureBuilder builder = ClusterFixture.builder(dirTestWatcher);
+    try (ClusterFixture cluster = builder.build();) {
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/test/java/org/apache/drill/exec/store/TestPluginRegistry.java]:
```suggestion
    try (ClusterFixture cluster = builder.build()) {
```

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginMap.java [hunk 2]
```diff
@@ -222,7 +222,9 @@ public synchronized PluginHandle remove(String name, StoragePluginConfig oldConf
         .build(logger);
     }
     nameMap.remove(oldEntry.name());
-    configMap.remove(oldEntry.config());
+    if (configMap.remove(oldEntry.config()) != oldEntry) {
+      throw new IllegalStateException("Config entry was modified while in the plugin cache");
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginMap.java]: Please add at least plugin name to make exception more informative.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistry.java [hunk 3]
```diff
@@ -91,14 +121,35 @@
...
+   * Safe way to add or remove a format plugin config from a stored file
+   * system configuration. Makes a copy of the config, adds/removes the
+   * format plugin, and updates the persistent store with the copy.
+   *
+   * @param pluginName name of the file system storage plugin config to
+   * modify
+   * @param formatName name of the format plugin to modify
+   * @param formatConfig if null, removes the plugin, if non-null updates
+   * the format plugin config with this value
+   * @throws ExecutionSetupException if the stoage plugin is undefined or
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistry.java]:
```suggestion
   * @throws ExecutionSetupException if the storage plugin is undefined or
```

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistryImpl.java [hunk 4]
```diff
@@ -396,41 +407,68 @@ public void put(String name, StoragePluginConfig config) throws ExecutionSetupEx
...
+    }
+  }
 
-      // Transfer the instance to a new handle, then invalidate the
-      // cache entry. The transfer ensures that the invalidate will
-      // not close the plugin instance
-      PluginHandle newHandle = ephemeralEntry.transfer(PluginType.STORED);
-      ephemeralPlugins.invalidate(config);
-      return newHandle;
+  // Configs are obtained from the persistent store. This method is
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistryImpl.java]: Please add text in form of java-doc.

**Comment by arina-ielchiieva** [comment]
@paul-rogers please add Jira title with link to Jira in PR description, since it's helps a lot to navigate between PR and Jira. Thanks!

**Comment by agozhiy** [comment]
Found a case when a plugin fails to be enabled:
1. On the storage page, disable a plugin (e.g. "dfs") using "disable" button
2. Update the plugin and manually set "enabled" option to true
3. Click Update

Expected result:
The plugin should be enabled.

Actual result:
It is still disabled.

**Comment by paul-rogers** [comment]
@agozhiy, thanks for your testing: you found a good corner case. Found and fixed the issue.

@arina, thanks for your review. Sorry, this commit adds a bunch more changes to clean up error handling, UI internals, etc.

Reran all unit tests. Verified the manual steps from @agozhiy, description.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistry.java [hunk 5]
```diff
@@ -145,16 +216,42 @@ void putFormatPlugin(String pluginName, String formatName,
...
    * Get the Schema factory associated with this storage plugin registry.
    *
    * @return A SchemaFactory that can register the schemas associated with this plugin registry.
    */
   SchemaFactory getSchemaFactory();
+
+  /**
+   * Object mapper to read/write the JSON form of a plugin
+   * config.
+   * @return
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistry.java]: Add description.

**Comment by arina-ielchiieva** [comment]
@paul-rogers I'll postpone the review till fix for the DRILL-7624 will be included.

**Comment by paul-rogers** [comment]
Updated to include fixes for [DRILL-7620](https://issues.apache.org/jira/browse/DRILL-7624).

Also, out of an abundance of caution, restored two method changes to avoid breaking third-party plugins. (Plugins in Drill were previously updated to use the new methods.)

Added tests to reproduce the issue in DRILL-7620 using a "known bad" mock plugin. Reran all unit tests, which passed.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistryImpl.java [hunk 6]
```diff
@@ -386,6 +387,20 @@ public void put(String name, StoragePluginConfig config) throws PluginException
...
 
+  @Override
+  public void validatedPut(String name, StoragePluginConfig config)
+      throws PluginException {
+
+    PluginHandle entry = createPluginEntry(name, config, PluginType.STORED);
+    try {
+      entry.plugin();
+    } catch (Exception e) {
+      throw new PluginException("Invalid plugin config", e);
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistryImpl.java]: Should we add plugin name to the error message?

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistry.java [hunk 7]
```diff
@@ -70,6 +109,27 @@
...
+  /**
+   * Return a config encoded as JSON.
+   * @throws PluginException if the plugin is undefined
+   */
+  String encode(String name) throws PluginException;
+  String encode(StoragePluginConfig config);
+
+  /**
+   * Return a config decoded from JSON.
+   * @throws IOException if the JSON is invalid
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistry.java]: PluginEncodingException?

**Comment by arina-ielchiieva** [comment]
@paul-rogers looks good, please address two minor comments and squash the commits.
@agozhiy could you please verify that PR addresses all found issues?

**Comment by paul-rogers** [comment]
Failed in 1.8. Seems a spurious error:

```
[ERROR] Failures: 
[ERROR]   TestLargeFileCompilation.testHashJoin:226->BaseTestQuery.resetSessionOption:380 Failed to reset session option `planner.enable_mergejoin`, Error: org.apache.drill.common.exceptions.UserException: SYSTEM ERROR: IllegalArgumentException: Attempted to send a message when connection is no longer valid.
```

**Comment by paul-rogers** [comment]
Thanks to everyone for their help on this. Sorry for the extra work as I fought with some of the more obscure corners of our plugin registry.

Addressed final comments. Squashed commits. Rebased on master.

**Comment by agozhiy** [comment]
@paul-rogers, I tested your recent changes to this PR and found some issues:
I used drill-embedded with previously wiped plugins (rm -rf /tmp/drill/sys.storage_plugins) to make them default.
Then:
1. Updated hive plugin to make it incorrect and enable it:
```json
{
  "type": "hive",
  "configProps": {
    "hive.metastore.sasl.enabled": "false"
  },
  "enabled": true
}
```
2. Tried to run a query with cp plugin:
```sql
SELECT * FROM cp.`employee.json` LIMIT 20
```
3. It worked, but a hive-related exception appeared in the logs (in case of embedded Drill it is being written directly to out)
4. Disabled the incorrect plugin and repeat the query to cp.
5. Observed that the exception happened again in spite of disabled hive plugin
Also once my query to cp hanged with the above exception looped, though I wasn't able to reproduce this again. 

I think, maybe it's better to return to the old behavior, when the it wasn't possible to enable plugin with incorrect config? And then fix it more thoroughly in a separate commit...

**Comment by arina-ielchiieva** [comment]
@paul-rogers I guess we should to address issues mentioned by Anton before moving forward...

**Comment by paul-rogers** [comment]
@agozhiy, here is how I read the old behavior:

* The UI sends the plugin as JSON via the `createOrUpdatePlugin()` mehod.
* `StorageResources` configures its mapper (which is really the global one from `LogicalPlanPersistence`) to accept comments, then deserializes the JSON to create a config.
* If deserialization fails, the UI returns an error message.
* `StorageResources` passes the name and config, as a `PluginConfigWrapper` to `createOrUpdatePluginJSON()` which calls `createOrUpdate()` in the plugin registry.
* `createOrUpdate()` first creates an instance of the plugin. The result will be one of a) `null` if the config is disabled, b) a `StoragePlugin` if enabled and valid, or c) an exception if invalid.
* `createOrUpdate()` now does one of:
  * If enabled, replace existing plugin in the in-memory cache with the new one, or add the new one
  * If disabled, remove the existing plugin from the in-memory cache
* `createOrUpdate()` then writes the config to persistent storage independent of the above choices.

There are bugs in the above. The above writes to the local cache before storage meaning that, for a brief time, the two are out of sync. If, during that time, another update comes in, then the two will again be out of sync. These issues don't crop up in the current tests, of course.

The new code does:

* The UI sends the plugin as JSON via the `createOrUpdatePlugin()` mehod.
* `StorageResources` calls `putJson()` in the plugin registry with the name and JSON text.
* `putJson()` decodes the JSON using the system mapper, configured to allow comments, throwing an exception if the decode fails. If succeeds, calls `put()` with the config.
* `put()` validates that the plugin is not a system plugin, then updates persistent storage.

I see the problem, the `put()` call above should be the new `validatedPut()` which creates an instance of the plugin first. Once we make that fix:

* `validatedPut()` creates an instance of the plugin, throwing an exception if instance creation fails.
* If OK, `validatedPut()` writes he plugin to persistent storage and updates the plugin cache.

There is a second bug here: the logic should be:

*  `validatedPut()` does one of:
   * If enabled, updates the config in the plugin cache.
   * If disabled, removes the config from the plugin cache.
* `validatedPut()` then checks if the old entry had a plugin instance. If , adds it to the ephemeral store.
* `validatedPut()`  writes the updated config to persistent storage.

The planner bugs explained earlier still occur. (A query should not iterate over all registered plugins.) Still, the above fixes should get us back to the original behavior, with the addition of the other fixes we're trying to make.

Made the two fixes noted above. Added unit tests for this case.

@agozhiy, if you run the UI scenario from your earlier comment, you should see that the plugin is rejected if invalid and enabled. The config is stored, but not instantiated, if invalid and disabled.

If you are brave, you can use ZK to change a stored valid config to make it invalid. Doing so simulates the external system failing. With the old version, Drill will fail on startup. With this version, Drill will start, but you'll see error messages in the log each time you plan a query. A future PR will ensure you see no errors at all.

All of this points out more open issues:

* If a plugin ever fails, should it be disabled? How would this work for a temporary outage? The user would have to manually re-enable the plugin when it becomes valid again.
* On a completely unrelated topic, there is code in the original version that forces all names of stored plugins to lower case (removing any mixed-case versions and replacing them), but no code enforces the lower-case rule in the in-memory cache. (I have tests that specifically test this case.) However, names are forced to lower case in storage. This seems more on the "bug" side of the ledger than the "feature" side. Should we allow mixed case everywhere or nowhere? Yet another thing to fix in the next PR.

**Comment by paul-rogers** [comment]
Rebased on latest master to resolve conflict.

**Comment by arina-ielchiieva** [comment]
@paul-rogers 

> if you run the UI scenario from your earlier comment, you should see that the plugin is rejected if invalid and enabled. The config is stored, but not instantiated, if invalid and disabled.

I like this behavior, +1 for this approach.

> If a plugin ever fails, should it be disabled? How would this work for a temporary outage? The user would have to manually re-enable the plugin when it becomes valid again.

I think in this case we should NOT disable the plugin. Once DB will be back online, assume we have rdbms plugin, user will be able to query using this plugin and no additional actions will be required. The problem in disabling is that only admin has permission to enable  / disable plugins, so on Production environments real users would first need to know when DB is back online and then ask admins to enable the plugin, which maybe not really convenient.

> On a completely unrelated topic, there is code in the original version that forces all names of stored plugins to lower case (removing any mixed-case versions and replacing them), but no code enforces the lower-case rule in the in-memory cache. (I have tests that specifically test this case.) However, names are forced to lower case in storage. This seems more on the "bug" side of the ledger than the "feature" side. Should we allow mixed case everywhere or nowhere? Yet another thing to fix in the next PR.

We should enforce lower-case plugin names everywhere.

**Comment by agozhiy** [comment]
@paul-rogers, the problem is: if a plugin is invalid and **disabled**, it is not being rejected. And later you can enable the plugin, causing the issue I described. (Validation should happen during enabling plugins).

**Comment by paul-rogers** [comment]
@agozhiy, thanks for finding the enable case.

Added validation on enabling the plugin. To do this right, needed to copy the config (can't change the state of a config already in the in-memory cache.) Making a copy uncovered `equals()` bugs in some format plugins, so fixed those.

Forced all plugin names to lower case.

Rebased on the latest master to resolve conflicts.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistryImpl.java [hunk 8]
```diff
@@ -233,6 +233,17 @@ private void loadIntrinsicPlugins() throws PluginException {
...
       for (StoragePlugin sysPlugin : intrinsicPlugins) {
+        // Enforce lower case names. Since the name of a system plugin
+        // is "hard coded", we can't adjust it if it is not already
+        // lower case. All we can do is fail to tell the developer that
+        // something is wrong.
+        String origName = sysPlugin.getName();
+        String lcName = sysPlugin.getName().toLowerCase();
+        if (!origName.equals(lcName)) {
+          throw new IllegalStateException(String.format(
+              "Plugin names must be lower case. System plugin name is not lower case: %s",
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistryImpl.java]:
```suggestion
              "Plugin names must be in lower case. System plugin name is not lower case: %s",
```

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistryImpl.java [hunk 9]
```diff
@@ -505,14 +530,19 @@ public StoragePluginConfig copyConfig(StoragePluginConfig orig) {
 
   @Override
   public StoragePluginConfig getDefinedConfig(String name) {
+    try {
+      name = validateName(name);
+    } catch (PluginException e) {
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistryImpl.java]: Should we log it somewhere or just ignore?

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/sequencefile/SequenceFileFormatConfig.java [hunk 10]
```diff
@@ -20,35 +20,42 @@
...
+    return Objects.equals(extensions, other.extensions);
+  }
+
+  @Override
+  public String toString() {
+    return new PlanStringBuilder(this)
+        .field("extensions", extensions)
+        .toString();
   }
 }
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/sequencefile/SequenceFileFormatConfig.java]: Please add new line

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistry.java [hunk 11]
```diff
@@ -96,6 +96,23 @@ public PluginEncodingException(String msg, Exception e) {
...
+   * store. If a formerly disabled plugin is enabled, verifies that the plugin
+   * can be instantiated as for {@link #verifiedPut()}.
+   * <p>
+   * Use this method when changing state. Do not obtain the config and change
+   * the state directly, doing so will make the plugin config inconsistent
+   * with the internal state.
+   *
+   * @param name name of the plugin
+   * @param enabled {@code true} to enable the plugin, {@code false} to disable
+   * @throws PluginNotFoundException it the plugin name is not valid
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistry.java]: Can't this exception be thrown is plugin is missing? Why only invalid name?

**Comment by arina-ielchiieva** [comment]
+1, LGTM
